src/main/java/at/ac/tuwien/kr/alpha/common/program/impl/InputProgram.java

```diff
@@ -0,0 +1,137 @@
+ */
+package at.ac.tuwien.kr.alpha.common.program.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import at.ac.tuwien.kr.alpha.common.atoms.AggregateLiteral;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.common.atoms.external.ExternalAtoms;
+import at.ac.tuwien.kr.alpha.common.program.AbstractProgram;
+import at.ac.tuwien.kr.alpha.common.rule.impl.BasicRule;
+import at.ac.tuwien.kr.alpha.grounder.parser.InlineDirectives;
+
+/**
+ * Alpha-internal representation of an ASP program, i.e., a set of ASP rules. Copyright (c) 2017 - 2019, the Alpha Team.
+ */
+public class InputProgram extends AbstractProgram<BasicRule> {
+
+	public static final InputProgram EMPTY = new InputProgram(Collections.emptyList(), Collections.emptyList(), new InlineDirectives());
+
+	public InputProgram(List<BasicRule> rules, List<Atom> facts, InlineDirectives inlineDirectives) {
+		super(rules, facts, inlineDirectives);
+	}
+
+	public InputProgram() {
+		super(new ArrayList<>(), new ArrayList<>(), new InlineDirectives());
+	}
+
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	public static Builder builder(InputProgram prog) {
+		return new Builder(prog);
+	}
+
+	/**
+	 * Builder to be used instead of accumulate(), ensuring that an @{link AbstractProgram} is immutable
+	 * 
+	 * Copyright (c) 2019, the Alpha Team.
+	 */
+	public static class Builder {
+
+		private List<BasicRule> rules = new ArrayList<>();
+		private List<Atom> facts = new ArrayList<>();
+		private InlineDirectives inlineDirectives = new InlineDirectives();
+
+		public Builder(InputProgram prog) {
+			this();
+			this.addRules(prog.getRules());
+			this.addFacts(prog.getFacts());
+			this.addInlineDirectives(prog.getInlineDirectives());
+		}
+
+		public Builder() {
+
+		}
+
+		public Builder addRules(List<BasicRule> rules) {
+			this.rules.addAll(rules);
+			return this;
+		}
+
+		public Builder addRule(BasicRule r) {
+			this.rules.add(r);
+			return this;
+		}
+
+		public Builder addFacts(List<Atom> facts) {
+			this.facts.addAll(facts);
+			return this;
+		}
+
+		public Builder addFact(Atom fact) {
+			this.facts.add(fact);
+			return this;
+		}
+
+		public <T extends Comparable<T>> Builder addExternalFacts(Class<T> clazz, Collection<T> extFacts) {
+			return this.addFacts(ExternalAtoms.asFacts(clazz, extFacts));
+		}
+
+		public Builder addInlineDirectives(InlineDirectives inlineDirectives) {
+			this.inlineDirectives.accumulate(inlineDirectives);
+			return this;
+		}
+
+		public Builder accumulate(InputProgram prog) {
+			return this.addRules(prog.getRules()).addFacts(prog.getFacts()).addInlineDirectives(prog.getInlineDirectives());
+		}
+
+		public InputProgram build() {
+			return new InputProgram(this.rules, this.facts, this.inlineDirectives);
+		}
+	}
+
+	public boolean containsAggregates() {
```

Comment:
Apparently this method is not used (any longer) and can be deleted.

Reply:
done